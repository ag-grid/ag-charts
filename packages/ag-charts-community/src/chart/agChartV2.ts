import type { LicenseManager } from '../module/enterpriseModule';
import { enterpriseModule } from '../module/enterpriseModule';
import { moduleRegistry } from '../module/module';
import { ChartOptions } from '../module/optionsModule';
import type { AgChartInstance, AgChartOptions, DownloadOptions, ImageDataUrlOptions } from '../options/agChartOptions';
import { Debug } from '../util/debug';
import { deepClone, jsonWalk } from '../util/json';
import { Logger } from '../util/logger';
import { mergeDefaults } from '../util/object';
import type { DeepPartial } from '../util/types';
import { VERSION } from '../version';
import { CartesianChart } from './cartesianChart';
import { Chart, type ChartExtendedOptions } from './chart';
import { AgChartInstanceProxy } from './chartProxy';
import { ChartUpdateType } from './chartUpdateType';
import { registerInbuiltModules } from './factory/registerInbuiltModules';
import { setupModules } from './factory/setupModules';
import { HierarchyChart } from './hierarchyChart';
import {
    isAgCartesianChartOptions,
    isAgHierarchyChartOptions,
    isAgPolarChartOptions,
    isAgTopologyChartOptions,
} from './mapping/types';
import { MementoCaretaker } from './memento';
import { PolarChart } from './polarChart';
import { TopologyChart } from './topologyChart';

const debug = Debug.create(true, 'opts');

function chartType(options: any): 'cartesian' | 'polar' | 'hierarchy' | 'topology' {
    if (isAgCartesianChartOptions(options)) {
        return 'cartesian';
    } else if (isAgPolarChartOptions(options)) {
        return 'polar';
    } else if (isAgHierarchyChartOptions(options)) {
        return 'hierarchy';
    } else if (isAgTopologyChartOptions(options)) {
        return 'topology';
    }

    throw new Error(`AG Chart - unknown type of chart for options with type: ${options.type}`);
}

/**
 * Factory for creating and updating instances of AgChartInstance.
 *
 * @docsInterface
 */
export abstract class AgCharts {
    private static readonly INVALID_CHART_REF_MESSAGE = 'AG Charts - invalid chart reference passed';
    private static licenseManager?: LicenseManager;
    private static licenseChecked = false;
    private static licenseKey?: string;
    private static gridContext = false;

    private static licenseCheck(options: AgChartOptions) {
        if (this.licenseChecked) return;

        this.licenseManager = enterpriseModule.licenseManager?.(options);
        this.licenseManager?.setLicenseKey(this.licenseKey, this.gridContext);
        this.licenseManager?.validateLicense();
        this.licenseChecked = true;
    }

    public static setLicenseKey(licenseKey: string) {
        this.licenseKey = licenseKey;
    }

    public static setGridContext(gridContext: boolean) {
        this.gridContext = gridContext;
    }

    public static getLicenseDetails(licenseKey: string) {
        return enterpriseModule.licenseManager?.({}).getLicenseDetails(licenseKey);
    }

    /**
     * Returns the `AgChartInstance` for a DOM node, if there is one.
     */
    public static getInstance(element: HTMLElement): AgChartInstance | undefined {
        return AgChartsInternal.getInstance(element);
    }

    /**
     * Create a new `AgChartInstance` based upon the given configuration options.
     */
    public static create(options: AgChartOptions): AgChartInstance {
        this.licenseCheck(options);
        const chart = AgChartsInternal.createOrUpdate(options);

        if (this.licenseManager?.isDisplayWatermark()) {
            enterpriseModule.injectWatermark?.(chart.chart.ctx.domManager, this.licenseManager.getWatermarkMessage());
        }
        return chart;
    }

    /**
     * Update an existing `AgChartInstance`. Options provided should be complete and not
     * partial.
     *
     * __NOTE__: As each call could trigger a chart redraw, multiple calls to update options in
     * quick succession could result in undesirable flickering, so callers should batch up and/or
     * debounce changes to avoid unintended partial update renderings.
     */
    public static update(chart: AgChartInstance, options: AgChartOptions) {
        if (!AgChartInstanceProxy.isInstance(chart)) {
            throw new Error(AgCharts.INVALID_CHART_REF_MESSAGE);
        }
        AgChartsInternal.createOrUpdate(options, chart);
    }

    /**
     * Update an existing `AgChartInstance` by applying a partial set of option changes.
     *
     * __NOTE__: As each call could trigger a chart redraw, each individual delta options update
     * should leave the chart in a valid options state. Also, multiple calls to update options in
     * quick succession could result in undesirable flickering, so callers should batch up and/or
     * debounce changes to avoid unintended partial update renderings.
     */
    public static updateDelta(chart: AgChartInstance, deltaOptions: DeepPartial<AgChartOptions>) {
        if (!AgChartInstanceProxy.isInstance(chart)) {
            throw new Error(AgCharts.INVALID_CHART_REF_MESSAGE);
        }
        AgChartsInternal.updateUserDelta(chart, deltaOptions);
    }

    /**
     * Starts a browser-based image download for the given `AgChartInstance`.
     */
    public static download(chart: AgChartInstance, options?: DownloadOptions) {
        if (!(chart instanceof AgChartInstanceProxy)) {
            throw new Error(AgCharts.INVALID_CHART_REF_MESSAGE);
        }
        AgChartsInternal.download(chart, options).catch((e) => Logger.errorOnce(e));
    }

    /**
     * Returns a base64-encoded image data URL for the given `AgChartInstance`.
     */
    public static getImageDataURL(chart: AgChartInstance, options?: ImageDataUrlOptions): Promise<string> {
        if (!(chart instanceof AgChartInstanceProxy)) {
            throw new Error(AgCharts.INVALID_CHART_REF_MESSAGE);
        }
        return AgChartsInternal.getImageDataURL(chart, options);
    }

    public static saveAnnotations(chart: AgChartInstance) {
        if (!(chart instanceof AgChartInstanceProxy)) {
            throw new Error(AgCharts.INVALID_CHART_REF_MESSAGE);
        }
        return AgChartsInternal.saveAnnotations(chart);
    }

    public static restoreAnnotations(chart: AgChartInstance, blob: unknown) {
        if (!(chart instanceof AgChartInstanceProxy)) {
            throw new Error(AgCharts.INVALID_CHART_REF_MESSAGE);
        }
        return AgChartsInternal.restoreAnnotations(chart, blob);
    }
}

class AgChartsInternal {
    private static readonly caretaker = new MementoCaretaker(VERSION);

    static getInstance(element: HTMLElement): AgChartInstanceProxy | undefined {
        const chart = Chart.getInstance(element);
        return chart ? AgChartInstanceProxy.chartInstances.get(chart) : undefined;
    }

    private static initialised = false;
    static initialiseModules() {
        if (AgChartsInternal.initialised) return;

        registerInbuiltModules();
        setupModules();

        AgChartsInternal.initialised = true;
    }

    static createOrUpdate(options: ChartExtendedOptions, proxy?: AgChartInstanceProxy) {
        AgChartsInternal.initialiseModules();

        debug('>>> AgChartV2.createOrUpdate() user options', options);

        const { overrideDevicePixelRatio, document, window: userWindow, ...userOptions } = options;
        const chartOptions = new ChartOptions(userOptions, { overrideDevicePixelRatio, document, window: userWindow });

        let chart = proxy?.chart;
        if (chart == null || chartType(userOptions) !== chartType(chart.processedOptions)) {
            chart = AgChartsInternal.createChartInstance(chartOptions, chart);
        }

        if (proxy == null) {
            proxy = new AgChartInstanceProxy(chart);
        } else {
            proxy.chart = chart;
        }

        if (debug.check() && typeof window !== 'undefined') {
            (window as any).agChartInstances ??= {};
            (window as any).agChartInstances[chart.id] = chart;
        }

        chart.queuedUserOptions.push(userOptions);
        chart.requestFactoryUpdate((chartRef) => {
            chartRef.applyOptions(chartOptions);
            // If there are a lot of update calls, `requestFactoryUpdate()` may skip callbacks,
            // so we need to remove all queue items up to the last successfully applied item.
            chartRef.queuedUserOptions.splice(0, chartRef.queuedUserOptions.indexOf(userOptions));
        });

        return proxy;
    }

    static updateUserDelta(proxy: AgChartInstanceProxy, deltaOptions: DeepPartial<AgChartOptions>) {
        deltaOptions = deepClone(deltaOptions, { shallow: ['data'] });

        jsonWalk(
            deltaOptions,
            (node) => {
                if (typeof node !== 'object') return;
                for (const [key, value] of Object.entries(node)) {
                    if (typeof value === 'undefined') {
                        Object.assign(node, { [key]: Symbol('UNSET') });
                    }
                }
            },
            { skip: ['data'] }
        );

        const { chart } = proxy;
        const lastUpdateOptions = chart.getOptions();
        const userOptions = mergeDefaults(deltaOptions, lastUpdateOptions);
        debug('>>> AgChartV2.updateUserDelta() user delta', deltaOptions);
        debug('AgChartV2.updateUserDelta() - base options', lastUpdateOptions);
        AgChartsInternal.createOrUpdate(userOptions, proxy);
    }

    /**
     * Returns the content of the current canvas as an image.
     */
    static async download(proxy: AgChartInstanceProxy, opts?: DownloadOptions) {
        try {
            const clone = await AgChartsInternal.prepareResizedChart(proxy, opts);
            clone.chart.ctx.scene.download(opts?.fileName, opts?.fileFormat);
            clone.destroy();
        } catch (error) {
            Logger.errorOnce(error);
        }
    }

    static async getImageDataURL(proxy: AgChartInstanceProxy, opts?: ImageDataUrlOptions): Promise<string> {
        const clone = await AgChartsInternal.prepareResizedChart(proxy, opts);
        const result = clone.chart.ctx.scene.getDataURL(opts?.fileFormat);

        clone.destroy();

        return result;
    }

    static saveAnnotations(proxy: AgChartInstanceProxy) {
        return AgChartsInternal.caretaker.save(proxy.chart.ctx.annotationManager);
    }

    static restoreAnnotations(proxy: AgChartInstanceProxy, blob: unknown) {
        return AgChartsInternal.caretaker.restore(proxy.chart.ctx.annotationManager, blob);
    }

    private static async prepareResizedChart({ chart }: AgChartInstanceProxy, opts: DownloadOptions = {}) {
        const width: number = opts.width ?? chart.width ?? chart.ctx.scene.canvas.width;
        const height: number = opts.height ?? chart.height ?? chart.ctx.scene.canvas.height;

        const options: ChartExtendedOptions = mergeDefaults(
            {
                container: document.createElement('div'),
                overrideDevicePixelRatio: 1,
                autoSize: false,
                width,
                height,
            },
            // Disable enterprise features that may interfere with image generation.
            moduleRegistry.hasEnterpriseModules() && { animation: { enabled: false } },
            chart.userOptions
        );

        const cloneProxy = AgChartsInternal.createOrUpdate(options);
        cloneProxy.chart.ctx.zoomManager.updateZoom('agChartV2', chart.ctx.zoomManager.getZoom()); // sync zoom
        chart.series.forEach((series, index) => {
            if (!series.visible) {
                cloneProxy.chart.series[index].visible = false; // sync series visibility
            }
        });
        chart.update(ChartUpdateType.FULL, { forceNodeDataRefresh: true });
        await cloneProxy.chart.waitForUpdate();
        return cloneProxy;
    }

    private static createChartInstance(options: ChartOptions, oldChart?: Chart): Chart {
        const transferableResource = oldChart?.destroy({ keepTransferableResources: true });
        const ChartConstructor = AgChartsInternal.getChartByOptions(options.processedOptions);
        return new ChartConstructor(options, transferableResource);
    }

    private static getChartByOptions(options: AgChartOptions) {
        if (isAgCartesianChartOptions(options)) {
            return CartesianChart;
        } else if (isAgHierarchyChartOptions(options)) {
            return HierarchyChart;
        } else if (isAgPolarChartOptions(options)) {
            return PolarChart;
        } else if (isAgTopologyChartOptions(options)) {
            return TopologyChart;
        }

        throw new Error(
            `AG Charts - couldn't apply configuration, check options are correctly structured and series types are specified`
        );
    }
}
