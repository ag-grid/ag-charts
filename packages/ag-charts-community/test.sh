#!/bin/bash

set -u

SCRIPT_DIR=$(cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd)
PROJECTS=""

if [[ $(basename $(pwd)) == "ag-charts-enterprise" ]] ; then
  FONTDIR=$PWD/src/test
else
  FONTDIR=$PWD/src/chart/test
fi

function testDev {
    npm run test:lint
    export FC_DEBUG=0
    export FONTCONFIG_PATH=$FONTDIR
    export FONTCONFIG_FILE=$FONTDIR/fonts.conf
    export PANGOCAIRO_BACKEND=fontconfig
    npx jest ${PROJECTS} --detectOpenHandles ${1:-}
}

function testCI {
    export FC_DEBUG=0
    export FONTCONFIG_PATH=$FONTDIR
    export FONTCONFIG_FILE=$FONTDIR/fonts.conf
    export PANGOCAIRO_BACKEND=fontconfig
    npx jest --runInBand
}

if [[ "$OSTYPE" != "linux-gnu"* ]] ; then
    echo -e "\033[41m\033[30mLinux is the only supported execution environment for Charts e2e tests.\033[0m"
    PROJECTS="--selectProjects unit"
fi

if [[ "${TEAMCITY_PROJECT_NAME:-}" == "" ]] ; then
  case ${1:-} in
    watch)
      testDev --watch
    ;;

    coverage)
      testDev --coverage
    ;;

    update)
      export SNAPSHOT_FAILURE_THRESHOLD=0.001
      testDev -u
    ;;

    *)
      testDev
    ;;
  esac
else
  testCI
fi
