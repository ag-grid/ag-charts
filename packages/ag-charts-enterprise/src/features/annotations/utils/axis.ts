import type { _ModuleSupport } from 'ag-charts-community';

export function calculateAxisLabelPadding(axisLayout: _ModuleSupport.AxisLayout) {
    return axisLayout.gridPadding + axisLayout.seriesAreaPadding + axisLayout.tickSize + axisLayout.label.padding;
}
