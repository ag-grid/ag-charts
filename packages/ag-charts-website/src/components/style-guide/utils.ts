export const formatName = (name) => {
    return name.replace('ag-grid', 'AG Grid').replaceAll('-', ' ');
};

export function hexToRGB(hex) {
    if (!hex) {
        return;
    }

    let r = 0,
        g = 0,
        b = 0;
    if (hex.length === 4) {
        r = parseInt('0x' + hex[1] + hex[1]);
        g = parseInt('0x' + hex[2] + hex[2]);
        b = parseInt('0x' + hex[3] + hex[3]);
    } else if (hex.length === 7) {
        r = parseInt('0x' + hex[1] + hex[2]);
        g = parseInt('0x' + hex[3] + hex[4]);
        b = parseInt('0x' + hex[5] + hex[6]);
    }

    return [r, g, b];
}

export function isLight(hex) {
    if (!hex) {
        return;
    }

    let [r, g, b] = hexToRGB(hex)!;

    return (r + g + b) / 3 > 150;
}

export function hexToHSL(hex) {
    if (!hex) {
        return;
    }

    // Convert hex to RGB first
    let [r, g, b] = hexToRGB(hex)!;

    // Then to HSL
    r /= 255;
    g /= 255;
    b /= 255;

    let cmin = Math.min(r, g, b),
        cmax = Math.max(r, g, b),
        delta = cmax - cmin,
        h = 0,
        s = 0,
        l = 0;

    if (delta == 0) {
        h = 0;
    } else if (cmax == r) {
        h = ((g - b) / delta) % 6;
    } else if (cmax == g) {
        h = (b - r) / delta + 2;
    } else {
        h = (r - g) / delta + 4;
    }

    h = Math.round(h * 60);

    if (h < 0) {
        h += 360;
    }

    l = (cmax + cmin) / 2;
    s = delta == 0 ? 0 : delta / (1 - Math.abs(2 * l - 1));
    s = +(s * 100).toFixed(1);
    l = +(l * 100).toFixed(1);

    return `hsl(${h}, ${s}%, ${l}%)`;
}
