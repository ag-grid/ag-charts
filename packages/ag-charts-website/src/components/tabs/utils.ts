import { load } from 'cheerio';

import { TAB_ID_PROP, TAB_LABEL_PROP } from './constants';
import type { TabData } from './types';

export function extractTabsData(html: string): TabData[] {
    const $ = load(html, null, false);

    return $(`[${TAB_LABEL_PROP}]`)
        .map((_index: number, node: any) => {
            const id = node.attribs[TAB_ID_PROP];
            const label = node.attribs[TAB_LABEL_PROP];
            const content = $(`[${TAB_LABEL_PROP}="${label}"]`).html() || '';

            return {
                id,
                label,
                content,
            };
        })
        .toArray();
}
