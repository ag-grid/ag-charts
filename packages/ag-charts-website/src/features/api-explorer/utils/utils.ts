export const formatJson = (object) => {
    const formatters = {
        number: (x) => x,
        string: (x) => `"${x}"`,
    };

    const formatArray = (array) => {
        if (array.length) {
            const type = typeof array[0];
            const format = formatters[type];

            if (format && array.every((x) => typeof x === type)) {
                return `[${array.map((x) => format(x)).join(', ')}]`;
            }
        }

        return array;
    };

    const formattedJson = JSON.stringify(object, (_, v) => (Array.isArray(v) ? formatArray(v) : v), 2);

    return formattedJson
        .replace(/'/g, "\\'") // escape single quotes
        .replace(/\\"/g, '"') // unescape double quotes
        .replace(/"([^"]+)":/g, '$1:') // format property names
        .replace(/"\[/g, '[')
        .replace(/\]"/g, ']') // finish array formatting
        .replace(/"([^"]+)"/g, "'$1'"); // force single quotes
};

export const deepClone = (object) => JSON.parse(JSON.stringify(object || {}));

export const getUrlParameters = () => {
    const parameters = {};

    window.location.search.replace(/[?&]+([^=&]+)=([^&]*)/gi, (m, key, value) => {
        parameters[key] = value;
    });

    return parameters;
};

export function isXAxisNumeric(chartType: string) {
    return ['scatter', 'histogram'].includes(chartType);
}
