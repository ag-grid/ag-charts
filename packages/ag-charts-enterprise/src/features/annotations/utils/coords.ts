import { _ModuleSupport, _Scene } from 'ag-charts-community';

import type { AnnotationContext, Point } from '../annotationTypes';
import { convertPoint, invertCoords } from './values';

const { Vec2 } = _ModuleSupport;
const { toRadians } = _Scene;

export function snapPoint(
    offset: _ModuleSupport.Vec2,
    context: AnnotationContext,
    snapping: boolean = false,
    origin?: Point,
    angleStep: number = 1
) {
    if (!snapping) return invertCoords(offset, context);

    const center = origin ? convertPoint(origin, context) : Vec2.origin();
    return invertCoords(snapToAngle(offset, center, angleStep), context);
}

export function snapToAngle(
    { x, y }: _ModuleSupport.Vec2,
    center: _ModuleSupport.Vec2,
    step: number,
    direction: number = 1
) {
    const { x: cx, y: cy } = center;
    const r = Math.sqrt(Math.pow(x - cx, 2) + Math.pow(y - cy, 2));
    const theta = Math.atan2(y - cy, x - cx);

    const stepRadians = toRadians(step);
    const snapTheta = Math.round(theta / stepRadians) * stepRadians;

    return {
        x: cx + r * Math.cos(snapTheta),
        y: cy + r * Math.sin(snapTheta) * direction,
    };
}
