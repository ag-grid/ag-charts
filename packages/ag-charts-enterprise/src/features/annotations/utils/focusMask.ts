import { _ModuleSupport, _Util } from 'ag-charts-community';

const { partialAssign, getWindow, createElement } = _ModuleSupport;
const { copyAttributes } = _Util;

export function createFocusMask(target: HTMLElement): HTMLElement {
    const mask = createElement('div');
    const targetStyle = getWindow().getComputedStyle(target);

    mask.className = 'ag-charts-focus-mask';
    mask.tabIndex = -1;
    mask.style.position = 'absolute';
    mask.style.pointerEvents = 'none';
    partialAssign(['width', 'height'], mask.style, targetStyle);
    copyAttributes(['role', 'aria-labelledby'], mask, target);

    const disappear = () => {
        mask.removeEventListener('keydown', disappear);
        mask.removeEventListener('blur', disappear);
        mask.remove();
    };
    mask.addEventListener('blur', disappear);
    mask.addEventListener('keydown', () => {
        target.focus();
        disappear();
    });

    target.insertAdjacentElement('beforebegin', mask);
    return mask;
}
