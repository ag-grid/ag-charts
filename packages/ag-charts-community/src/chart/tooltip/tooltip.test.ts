import { describe, expect, it } from '@jest/globals';
import { toMatchImageSnapshot } from 'jest-image-snapshot';

import type { AgChartOptions } from '../../options/agChartOptions';
import { AgCharts } from '../agChartV2';
import type { Chart } from '../chart';
import { expectWarning, setupMockConsole } from '../test/mockConsole';
import { AgChartProxy, createChart, hoverAction, prepareTestOptions, waitForChartStability } from '../test/utils';

expect.extend({ toMatchImageSnapshot });

describe('Tooltip', () => {
    setupMockConsole();

    describe('Validation', () => {
        it('should show 1 warning for invalid tooltip value', async () => {
            await createChart({
                data: [
                    { month: 'Jun', sweaters: 50 },
                    { month: 'Jul', sweaters: 70 },
                    { month: 'Aug', sweaters: 60 },
                ],
                series: [{ type: 'line', xKey: 'month', yKey: 'sweaters', yName: 'Sweaters Made' }],
                tooltip: {
                    position: '2' as any,
                },
            });

            expectWarning(
                `AG Charts - unable to set [tooltip.position] in Tooltip - can't apply type of [primitive], allowed types are: [class-instance]`
            );
        });

        it('should show 1 warning for invalid tooltip position value', async () => {
            await createChart({
                data: [
                    { month: 'Jun', sweaters: 50 },
                    { month: 'Jul', sweaters: 70 },
                    { month: 'Aug', sweaters: 60 },
                ],
                series: [{ type: 'line', xKey: 'month', yKey: 'sweaters', yName: 'Sweaters Made' }],
                tooltip: {
                    position: { type: 'ponter' as any, xOffset: 80, yOffset: 80 },
                },
            });

            expectWarning(
                `AG Charts - Property [type] of [TooltipPosition] cannot be set to ["ponter"]; expecting a position type keyword such as 'pointer', 'node', 'top', 'right', 'bottom', 'left', 'top-left', 'top-right', 'bottom-right' or 'bottom-left', ignoring.`
            );
        });
    });

    describe('Realtime', () => {
        let chart: AgChartProxy;
        afterEach(() => {
            chart?.destroy();
        });

        it('should update tooltip correctly', async () => {
            // See AG-10409: The tooltip should update when the mouse stays in place but the data is updated.
            const opts: AgChartOptions = prepareTestOptions({});
            opts.data = [
                { time: 0, voltage: 1.362460821419385 },
                { time: 1, voltage: 1.3072694395877953 },
                { time: 2, voltage: 1.1967684308904354 },
                { time: 3, voltage: 1.2362572382997417 },
                { time: 4, voltage: 1.479504186572628 },
                { time: 5, voltage: 1.401144010767596 },
                { time: 6, voltage: 1.2192725536972913 },
                { time: 7, voltage: 1.1097886105628154 },
                { time: 8, voltage: 1.4869931693640273 },
                { time: 9, voltage: 1.1720928254975662 },
            ];
            opts.series = [{ type: 'line', xKey: 'time', yKey: 'voltage' }];

            chart = AgCharts.create(opts) as AgChartProxy;
            await waitForChartStability(chart);

            const nextValue = async (time: number, voltage: number) => {
                opts.data!.shift();
                opts.data!.push({ time, voltage });
                AgCharts.update(chart, opts);
                await waitForChartStability(chart);
            };

            await hoverAction(400, 300)(chart);
            await waitForChartStability(chart);
            expect(document.body.getElementsByClassName('ag-chart-tooltip')).toMatchSnapshot();

            await nextValue(10, 1.3249187570726666);
            expect(document.body.getElementsByClassName('ag-chart-tooltip')).toMatchSnapshot();

            await nextValue(11, 1.2651169069335022);
            expect(document.body.getElementsByClassName('ag-chart-tooltip')).toMatchSnapshot();

            await nextValue(12, 1.3627720015958902);
            expect(document.body.getElementsByClassName('ag-chart-tooltip')).toMatchSnapshot();

            await nextValue(13, 1.490244608234256);
            expect(document.body.getElementsByClassName('ag-chart-tooltip')).toMatchSnapshot();

            await nextValue(14, 1.490244608234256);
            expect(document.body.getElementsByClassName('ag-chart-tooltip')).toMatchSnapshot();
        });
    });
});
