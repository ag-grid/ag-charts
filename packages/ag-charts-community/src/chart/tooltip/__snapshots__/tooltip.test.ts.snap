// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip Realtime should update tooltip correctly 1`] = `
HTMLCollection [
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-hidden ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
  />,
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-hidden ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
  />,
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
    style="transform: translate(365px, 40px);"
  >
    <div
      class="ag-chart-tooltip-content"
    >
      4: 1.48
    </div>
  </div>,
]
`;

exports[`Tooltip Realtime should update tooltip correctly 2`] = `
HTMLCollection [
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-hidden ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
  />,
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-hidden ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
  />,
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
    style="transform: translate(365px, 147px);"
  >
    <div
      class="ag-chart-tooltip-content"
    >
      5: 1.40
    </div>
  </div>,
]
`;

exports[`Tooltip Realtime should update tooltip correctly 3`] = `
HTMLCollection [
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-hidden ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
  />,
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-hidden ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
  />,
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
    style="transform: translate(365px, 397px);"
  >
    <div
      class="ag-chart-tooltip-content"
    >
      6: 1.22
    </div>
  </div>,
]
`;

exports[`Tooltip Realtime should update tooltip correctly 4`] = `
HTMLCollection [
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-hidden ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
  />,
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-hidden ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
  />,
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
    style="transform: translate(365px, 547px);"
  >
    <div
      class="ag-chart-tooltip-content"
    >
      7: 1.11
    </div>
  </div>,
]
`;

exports[`Tooltip Realtime should update tooltip correctly 5`] = `
HTMLCollection [
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-hidden ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
  />,
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-hidden ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
  />,
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
    style="transform: translate(365px, 30px);"
  >
    <div
      class="ag-chart-tooltip-content"
    >
      8: 1.49
    </div>
  </div>,
]
`;

exports[`Tooltip Realtime should update tooltip correctly 6`] = `
HTMLCollection [
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-hidden ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
  />,
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-hidden ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
  />,
  <div
    class="ag-chart-tooltip ag-chart-tooltip-no-interaction ag-chart-tooltip-arrow ag-chart-tooltip-wrap-hyphenate"
    style="transform: translate(365px, 461px);"
  >
    <div
      class="ag-chart-tooltip-content"
    >
      9: 1.17
    </div>
  </div>,
]
`;
